### PR TITLE
out_s3: add Apache Arrow support

### PR DESCRIPTION
**Pull request by fujimotos**
[Apache Arrow](https://arrow.apache.org/) is an efficient columnar data format that is suitable
for statistical analysis, and popular in machine learning community.

With this patch merged, users now can specify 'arrow' as the
compression type like this:

```ini
[OUTPUT]
  Name s3
  Bucket some-bucket
  total_file_size 1M
  use_put_object On
  compression arrow
```

which makes Fluent Bit convert the request buffer into Apache Arrow
format before uploading.

Signed-off-by: Fujimoto Seiji <fujimoto@clear-code.com>

### CODE REVIEW

**Comment by fujimotos** [comment]
Arrow support is disabled by default because not every server has the required
dependency `libarrow`.  To enable the support, we need to compile it as follows:

```sh
$ cmake .. -GFLB_ARROW=On
$ cmake --build .
```

Here is an example that shows how one can utilize Apache Arrow support.

### Configuration

```ini
[INPUT]
  Name cpu 

[OUTPUT]
  Name s3
  Match *
  Region ap-northeast-1
  Bucket fluent-bit-20210308
  total_file_size 1M
  use_put_object On
  upload_timeout 1m
  Compression arrow
```

### Result

Now the uploaded data can be loaded instantly via Arrow's S3 interface.

https://arrow.apache.org/docs/python/filesystems.html

For example, the above configuration produces a very clean tabular time-series
data like this:

```python3
>>> import pyarrow as pa
>>> table = load_data_from_s3()
>>> print(table)
                           date  cpu_p  user_p  system_p  cpu0.p_cpu  cpu0.p_user  cpu0.p_system
0   2021-03-08T09:03:03.668251Z    0.0     0.0       0.0         0.0          0.0            0.0 
1   2021-03-08T09:03:04.668156Z    1.0     1.0       0.0         1.0          1.0            0.0 
2   2021-03-08T09:03:05.668242Z    0.0     0.0       0.0         0.0          0.0            0.0 
3   2021-03-08T09:03:06.668269Z    0.0     0.0       0.0         0.0          0.0            0.0 
4   2021-03-08T09:03:07.668218Z    0.0     0.0       0.0         0.0          0.0            0.0 
5   2021-03-08T09:03:08.739886Z    2.0     1.0       1.0         2.0          1.0            1.0 
6   2021-03-08T09:03:09.668181Z    1.0     1.0       0.0         1.0          1.0            0.0 
7   2021-03-08T09:03:10.668247Z    1.0     0.0       1.0         1.0          0.0            1.0 
8   2021-03-08T09:03:11.668182Z    2.0     2.0       0.0         2.0          2.0            0.0 
9   2021-03-08T09:03:12.668275Z    1.0     0.0       1.0         1.0          0.0            1.0 
10  2021-03-08T09:03:13.668428Z    0.0     0.0       0.0         0.0          0.0            0.0 
11  2021-03-08T09:03:14.668320Z    2.0     2.0       0.0         2.0          2.0            0.0 
12  2021-03-08T09:03:15.668256Z    0.0     0.0       0.0         0.0          0.0            0.0 
13  2021-03-08T09:03:16.668287Z    0.0     0.0       0.0         0.0          0.0            0.0 
14  2021-03-08T09:03:17.668307Z    1.0     1.0       0.0         1.0          1.0            0.0 
15  2021-03-08T09:03:18.668257Z    0.0     0.0       0.0         0.0          0.0            0.0 
16  2021-03-08T09:03:19.668281Z    0.0     0.0       0.0         0.0          0.0            0.0 
17  2021-03-08T09:03:20.668317Z    0.0     0.0       0.0         0.0          0.0            0.0 
18  2021-03-08T09:03:21.668231Z    0.0     0.0       0.0         0.0          0.0            0.0 
19  2021-03-08T09:03:22.668222Z    0.0     0.0       0.0         0.0          0.0            0.0 
```

**Comment by PettitWesley** [comment]
CC @zhonghui12

File: plugins/out_s3/s3.c
Comment (zhonghui12) [inline plugins/out_s3/s3.c:144]: I think we should create a header for apache arrow so that users can know if the data is compressed from the log.

**Comment by agup006** [comment]
I'm wondering if we can just add the build requirements to the build server, seems like it could help with this feature adoption.

 I'm thinking of how the tensorflow filter is rarely used because a user needs to build with specific settings on in order to enable it.

**Review by PettitWesley — APPROVED** [review]
Aside from the question of gating this behind a build flag that defaults to false, the code LGTM

**Comment by fujimotos** [comment]
> I'm wondering if we can just add the build requirements to the build server, seems like it could help with this feature adoption.
> I'm thinking of how the tensorflow filter is rarely used because a user needs to build with specific settings on in order to enable it.

@kou Do you have any comments on this point?

The basic problem here is that enabling Arrow support makes Fluent Bit executable
dependent on `libarrow`, but I don't think most servers have the library installed.

**Comment by kou** [comment]
If we enable Apache Arrow support by default, we should use Apache Arrow C++ directly instead of using via Apache Arrow GLib. If we use Apache Arrow GLib, we need `libarrow`, `libarrow-glib` and `libglib`.
(And we can avoid performance overhead by Apache Arrow GLib.)

We can use `libarrow.a` with `ExternalProject_Add` and `-DARROW_BUILD_STATIC=ON` `CMAKE_ARGS` as fallback when users don't have installed `libarrow`.

File: plugins/out_s3/s3.c
Comment (fujimotos) [reply plugins/out_s3/s3.c]: Thank you for the comment. I fixed it in 9287d104.

File: plugins/out_s3/s3.c
Comment (fujimotos) [reply plugins/out_s3/s3.c]: Fixed.

File: plugins/out_s3/s3.c
Comment (fujimotos) [reply plugins/out_s3/s3.c]: Fixed.

**Comment by fujimotos** [comment]
@edsiper I updated this patch accordingly.

BTW, I  discussed with @kou about enabling Apache Arrow support by default:
It is difficult, as of Mar 2021, because `libarrow` is implemented in C++, and we
cannot easily statically link it against Fluent Bit (which is C).

There is [a future plan to create a C-friendly library within Apache Arrow project](https://issues.apache.org/jira/browse/ARROW-1851).
Until that statically-linkable library is available, the current implementation is
the best we can do.

File: plugins/out_s3/arrow/compress.c
Comment (kou) [inline plugins/out_s3/arrow/compress.c:101]: Apache Arrow has two file formats:

1. Streaming format: https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format
2. File format: https://arrow.apache.org/docs/format/Columnar.html#ipc-file-format

This code uses the streaming format. The streaming format is suitable for stream data. Users can start processing before they receive all data. But users can't do random access.

The file format is suitable for file data. Users can't start processing before they receive all data. But they can do random access.

I think that file format is suitable for data on S3. Users can do random access by HTTP range header without downloading all data on S3.

We can use convenient `garrow_table_write_as_feather()` for the file format. Note that "Feather v2" is another name of the file format: https://arrow.apache.org/docs/python/feather.html

```c
success = garrow_table_write_as_feather(table, GARROW_OUTPUT_STREAM(sink), NULL, &errors);
```

We can compress data by LZ4 or Zstandard:

```c
GArrowFeatherWriteProperties *properties = garrow_feather_write_properties_new();
g_object_set(G_OBJECT(properties),
            "compression", GARROW_COMPRESSION_TYPE_ZSTD,
            NULL);
success = garrow_table_write_as_feather(table, GARROW_OUTPUT_STREAM(sink), properties, &errors);
g_object_unref(properties);
```

File: CMakeLists.txt
Comment (fujimotos) [reply CMakeLists.txt]: Fixed.

File: plugins/out_s3/arrow/compress.c
Comment (fujimotos) [reply plugins/out_s3/arrow/compress.c]: Thank you. I changed `compress.c` to use Feather v2 format with ZSTD compression.

**Comment by fujimotos** [comment]
@kou I applied your feedback. Please approve this PR if you are fine.

File: plugins/out_s3/arrow/compress.c
Comment (kou) [inline plugins/out_s3/arrow/compress.c:102]: Zstandard support is optional in Apache Arrow.
If Apache Arrow isn't built with Zstandard support, `garrow_table_write_feather()` is always failed.

It's better that compression is configurable.

File: plugins/out_s3/arrow/compress.c
Comment (kou) [inline plugins/out_s3/arrow/compress.c:15]: I think that "data frame" is better than "matrix".
It seems that "matrix" uses the same type for all elements but "data frame" may use different type for each column.

**Review by kou — DISMISSED** [review]

**Comment by PettitWesley** [comment]
@fujimotos what is needed to get this feature finished?

**Comment by fujimotos** [comment]
@PettitWesley Sorry. I've been a bit absent from Fluent Bit recently,
working on Fluentd-side of stuffs (I'm also a maintainer of that project).

I'm gonna find some time today to finish this PR! So WFM.

**Comment by fujimotos** [comment]
I took some time today updating this PR.

* Rebase the change set to master HEAD to resolve the conflict.
* Remove ZSTD compression support (feedback from kou)
* Fix "matrix" to "dataframe" in the comment (feedback from oku)

I can confirm that Fluent Bit can upload Apache Arrow data to S3 (with fdb7147)

**Comment by fujimotos** [comment]
@kou Can you check this PR again? I updated this PR to incorporate
your feedback.

If you're fine, I think this PR can be merged into mainline.

File: plugins/out_s3/arrow/compress.c
Comment (fujimotos) [reply plugins/out_s3/arrow/compress.c]: Fixed.

File: plugins/out_s3/arrow/compress.c
Comment (fujimotos) [reply plugins/out_s3/arrow/compress.c]: Fixed.

**Review by kou — APPROVED** [review]
It looks good to me with a small fix.

Could you also replace `ctx->compression != NULL` with `ctx->compression != 0` here?
https://github.com/fluent/fluent-bit/blob/master/plugins/out_s3/s3.c#L1131

**Comment by fujimotos** [comment]
> Could you also replace ctx->compression != NULL with ctx->compression != 0 here?

@kou Thank you! I have fixed that.

**Review by kou — APPROVED** [review]
+1

**Comment by fujimotos** [comment]
@PettitWesley I believe this PR is mergeable now.

**Comment by fujimotos** [comment]
fluent/fluent-bit-docs/pull/523 is the documentation patch for the feature.

**Comment by PettitWesley** [comment]
@fujimotos Awesome!

@edsiper You requested changes; can you approve/re-review so we can get this merged?

File: plugins/out_s3/arrow/compress.c
Comment (edsiper) [inline plugins/out_s3/arrow/compress.c:39]: plugin errors should use flb_plg_abc() familfy functions

**Review by edsiper — CHANGES_REQUESTED** [review]

**Comment by PettitWesley** [comment]
@edsiper @fujimotos What do we need to get this merged?

**Comment by edsiper** [comment]
@PettitWesley 

- get rid of g_print() function (replace by flb_plg...()) 
- fix conflicts
- @PettitWesley you approve the changes once listed items above are fixed

**Comment by fujimotos** [comment]
@PettitWesley @edsiper Sorry for being late. I submit a update e210a45.

> * get rid of g_print() function

I cleaned up `g_print()` functions from my implementation. 

> * fix conflicts

It's now cleanly mergiable with master.

**Comment by fujimotos** [comment]
ADDENDUM: I can confirm e210a45 fine with the latest version of Apache Arrow (4.0)
So this patch should be good in terms of features.

**Review by PettitWesley — APPROVED** [review]
@fujimotos Thanks!

**Comment by fujimotos** [comment]
I merged this PR into mainline via 544fa896a5c362a979cd251deb990061b9c2aa96.

Thanks for all who reviewed and helped this PR.  Close this PR now.

**Pull request closed without merging**
